### PR TITLE
SI-6528 Avoid loop in getClassParts.

### PR DESCRIPTION
**Pull request by retronym**
Carefully crafted types can lead to an endless stream of
distinct parts, differing only in different instantiations
of an abstract type.

This commit substitutes abstract types with their upper bounds
before determining if a type has already been seen.

Review for correctness, completeness, and performance by @adriaanm.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -982,9 +982,15 @@ trait Implicits {
        *                  in order to cache them in infoMapCache
        */
       def getParts(tp: Type)(implicit infoMap: InfoMap, seen: mutable.Set[Type], pending: Set[Symbol]) {
-        if (seen(tp))
+        // SI-6528 Avoid following an endless stream of types that differ only in
+        //         abstract types
+        val seenType = tp.map {
+          case tp @ TypeRef(_, sym, args) if sym.isAbstractType => tp.bounds.hi
+          case t => t
+        }
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:990]: Compare with abstractTypeToBounds[1], which has several more cases. Can you elaborate on the difference? Intuitively, if one wants this case, one wants the rest of them.

[1] https://github.com/scala/scala/blob/78bcc844f1/src/reflect/scala/reflect/internal/Types.scala#L3709

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:990]: I didn't find that one. I'll see if I can generate variations on the test to motivate it, but even if I can't it looks preferable.

**Comment by retronym** [comment]
Closed for renovations. I need to understand this a bit better. (Switching to `abstractTypeToBounds` broke `test/files/pos/t1203.scala`, related to annotated types.)

**Comment by paulp** [comment]
Since abstractTypeToBounds recurses through AnnotatedType, it's pretty believable that you'd witness different behavior; of interest to me is what happens if you change abstractTypeToBounds not to recurse through AnnotatedType. If there is no visible behavioral change (that's my opening hypothesis) it doesn't mean there's nothing useful taking place, but it does mean the usual thing that whatever the useful thing is, we have no test for it or documentation of it.

The AnnotatedType case was added in b907c8eb599b5 .

**Pull request closed without merging**